test/trento/domain/sap_system/sap_system_test.exs

```diff
@@ -4202,6 +4208,361 @@ defmodule Trento.SapSystemTest do
+          absent_at: DateTime.utc_now()
+        ),
+        build(:application_instance_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid,
+          host_id: host_id,
+          instance_number: present_app_instance_number,
+          features: "MESSAGESERVER"
+        ),
+        build(
+          :sap_system_registered_event,
+          sap_system_id: sap_system_id,
+          sid: sid,
+          ensa_version: ensa_version
+        )
+      ]
+
+      assert_events_and_state(
+        initial_events,
+        [
+          %RegisterDatabaseInstance{
+            sap_system_id: sap_system_id,
+            host_id: host_id,
+            instance_number: absent_db_instance_number,
+            health: :passing
+          },
+          %RegisterDatabaseInstance{
+            sap_system_id: sap_system_id,
+            host_id: host_id,
+            instance_number: present_db_instance_number,
+            health: :passing
+          },
+          %RegisterApplicationInstance{
+            sap_system_id: sap_system_id,
+            host_id: host_id,
+            instance_number: absent_app_instance_number,
+            health: :passing,
+            ensa_version: ensa_version
+          },
+          %RegisterApplicationInstance{
+            sap_system_id: sap_system_id,
+            host_id: host_id,
+            instance_number: present_app_instance_number,
+            health: :passing,
+            ensa_version: ensa_version
+          }
+        ],
+        [
+          %DatabaseInstanceMarkedPresent{
+            instance_number: absent_db_instance_number,
+            host_id: host_id,
+            sap_system_id: sap_system_id
+          },
+          %ApplicationInstanceMarkedPresent{
+            instance_number: absent_app_instance_number,
+            host_id: host_id,
+            sap_system_id: sap_system_id
+          }
+        ],
+        fn state ->
+          assert %SapSystem{
+                   sid: ^sid,
+                   ensa_version: ^ensa_version,
+                   application: %SapSystem.Application{
+                     sid: ^sid,
+                     instances: [
+                       %SapSystem.Instance{
+                         absent_at: nil
+                       },
+                       %SapSystem.Instance{
+                         absent_at: nil
+                       }
+                     ]
+                   },
+                   database: %SapSystem.Database{
+                     sid: ^sid,
+                     instances: [
+                       %SapSystem.Instance{
+                         absent_at: nil
+                       },
+                       %SapSystem.Instance{
+                         absent_at: nil
+                       }
+                     ]
+                   }
+                 } = state
+        end
+      )
+    end
+
+    test "absent instances receiving 'mark as absent' command should mark as absent" do
```

Comment:
I would put this test before the others.
The tests can tell the story of the flow as well, so we would set to absent before setting it back to present, right?